### PR TITLE
feat: rspack-provider support tools.babel

### PR DESCRIPTION
**Pull request by GiveMe-A-Name**
## Description

In some times, user maybe need add custom babel plugin or babel presets. 
`rspack-provider ` need support it

### UNSAFE

1. should we need add core-js for every entries in rspack?

2. the rspack is not support includes is a object. So It is a unsafe action.
```ts
  // compile all folders in app directory, exclude node_modules
  rule.include.add(context.rootPath);

  // it will let user can't include the package in `node_modules`.
  rule.exclude.add(/node_modules/);

  // ....
``` 

## Types of changes

- [ ] Docs change / Dependency upgrade
- [ ] Bug fix
- [x] New feature / Improvement
- [ ] Refactoring
- [ ] Breaking change

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/builder/builder-shared/package.json [hunk 1]
```diff
@@ -67,6 +68,7 @@
     }
   },
   "dependencies": {
+    "@modern-js/babel-preset-app": "workspace:*",
```
Comment (chenjiahan) [inline packages/builder/builder-shared/package.json:71]: Most Rspack projects do not need babel, if we declare the "@modern-js/babel-preset-app" dependency in `builder-shared` by default, the install size will be bigger.

File: packages/builder/builder-rspack-provider/src/plugins/babel.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      rule
+        .test(mergeRegex(JS_REGEX, TS_REGEX))
+        .use(CHAIN_ID.USE.BABEL)
+        .loader(getSharedPkgCompiledPath('babel-loader'))
+        .options(babelOptions);
+    });
+  },
+});
+
+export function applyScriptCondition({
```
Comment (9aoy) [inline packages/builder/builder-rspack-provider/src/plugins/babel.ts:78]: 这个方法可以抽到 shared 里吧？

File: packages/builder/builder-webpack-provider/src/plugins/babel.ts [hunk 3]
```diff
@@ -170,7 +164,7 @@ export const builderPluginBabel = (): BuilderPlugin => ({
         rule
           .test(useTsLoader ? JS_REGEX : mergeRegex(JS_REGEX, TS_REGEX))
           .use(CHAIN_ID.USE.BABEL)
-          .loader(getCompiledPath('babel-loader'))
+          .loader(getSharedPkgCompiledPath('babel-loader'))
```
Comment (9aoy) [inline packages/builder/builder-webpack-provider/src/plugins/babel.ts:167]: prebundle 的脚本不需要改么？  webpack-provider 里的 babel-loader bundle 也要移除吧

File: packages/builder/builder-rspack-provider/src/plugins/babel.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        context: api.context,
+        includes,
+        excludes,
+      });
+
+      rule
+        .test(mergeRegex(JS_REGEX, TS_REGEX))
+        .use(CHAIN_ID.USE.BABEL)
+        .loader(getSharedPkgCompiledPath('babel-loader'))
+        .options(babelOptions);
```
Comment (9aoy) [inline packages/builder/builder-rspack-provider/src/plugins/babel.ts:73]: e2e 测试补充一个吧

File: packages/builder/builder-webpack-provider/src/plugins/babel.ts (hunk 3 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-webpack-provider/src/plugins/babel.ts:140]: webpack-provider 里的 babel-loader bundle 移除了

File: packages/builder/builder-rspack-provider/src/plugins/babel.ts (hunk 4 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-rspack-provider/src/plugins/babel.ts:73]: Got it

File: packages/builder/builder-shared/package.json (hunk 1 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-shared/package.json:71]: Maybe in long time, the app-tools would default dependent the `webpack-provider`, It is mean we already dependent the `@modern-js/babel-preset-app` by default. 

That's a problem exactly, but it 's not a problem now.
We should resolve this when we switch the bundler engine to rspack by default

File: packages/builder/builder-rspack-provider/src/plugins/babel.ts (hunk 2 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-rspack-provider/src/plugins/babel.ts]: done

File: packages/builder/builder-shared/package.json (hunk 1 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-shared/package.json:71]: emm. Maybe we need think carefully about this problem.

**Review by 9aoy — DISMISSED** [review]

File: packages/builder/builder-rspack-provider/src/plugins/babel.ts (hunk 4 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-rspack-provider/src/plugins/babel.ts:85]: Done~

File: packages/builder/builder-shared/package.json (hunk 1 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-shared/package.json]: Updated it.`@modern-js/babel-preset-app` would be a dev dependency in `builder-shared`.

but the `babel-loader` still in `builder-shared`, There are two main reasons follow: 
- `babel-loader` only 242 kb size.
- The preBuild dependency does not load well dynamically

**Review by chenjiahan — APPROVED** [review]

**Review by 9aoy — APPROVED** [review]

**Pull request merged**